python/cog/server/worker.py

```diff
@@ -360,8 +403,46 @@ def _predict(
                 else:
                     self._events.send(PredictionOutputType(multi=False))
                     self._events.send(PredictionOutput(payload=make_encodeable(result)))
+
+    async def _apredict(
+        self,
+        payload: Dict[str, Any],
+        predict: Callable[..., Any],
+        redirector: AsyncStreamRedirector,
+    ) -> None:
+        with self._handle_predict_error(redirector):
+            result = predict(**payload)
+
+            if result:
+                if inspect.isasyncgen(result):
+                    self._events.send(PredictionOutputType(multi=True))
+                    async for r in result:
+                        self._events.send(PredictionOutput(payload=make_encodeable(r)))
+                else:
+                    output = await result
+                    self._events.send(PredictionOutputType(multi=False))
+                    self._events.send(PredictionOutput(payload=make_encodeable(output)))
+
+    @contextlib.contextmanager
+    def _handle_predict_error(
+        self, redirector: Union[AsyncStreamRedirector, StreamRedirector]
+    ) -> Iterator[None]:
+        done = Done()
+        send_done = True
+        self._cancelable = True
+        try:
+            yield
+        # regular cancelation
         except CancelationException:
             done.canceled = True
+        # async cancelation
+        except asyncio.CancelledError:
+            done.canceled = True
+            # We've handled the requested cancelation, so we uncancel the task.
+            # This ensures that any cleanup work we do won't be interrupted.
+            task = asyncio.current_task()
+            assert task
+            task.uncancel()
```

Comment:
whoa

I _think_ cleanup work isn't supposed to be interrupted, at least not in a way that's prevented by uncancelling, but it would be fantastic to test this

Reply:
I think you might be right that this isn't necessary. I found [this](https://github.com/python/cpython/issues/102780) which suggested to me that it was, but I think that was just a bug in the `asyncio.timeout` context manager.

Reply:
yeah, that doesn't seem like it's related to our use of TimeoutError, but that is interesting